vendor/bat-native-ledger/src/bat/ledger/internal/core/bat_ledger_context.cc

```diff
@@ -7,89 +7,101 @@
 
 #include <utility>
 
+#include "bat/ledger/internal/constants.h"
 #include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/ledger.h"
 #include "bat/ledger/ledger_client.h"
+#include "bat/ledger/option_keys.h"
+#include "bat/ledger/public/interfaces/ledger.mojom.h"
 
 namespace ledger {
 
 namespace {
 
-size_t g_next_component_key = 0;
+BATLedgerContext::Environment GetEnvironment() {
+  switch (ledger::_environment) {
+    case mojom::Environment::DEVELOPMENT:
+      return BATLedgerContext::Environment::kDevelopment;
+    case mojom::Environment::STAGING:
+      return BATLedgerContext::Environment::kStaging;
+    case mojom::Environment::PRODUCTION:
+      return BATLedgerContext::Environment::kProduction;
+  }
+}
+
+BATLedgerContext::Options GetOptions() {
+  return {.environment = GetEnvironment()};
+}
 
 }  // namespace
 
 BATLedgerContext::BATLedgerContext(LedgerImpl* ledger_impl)
-    : ledger_client_(ledger_impl->ledger_client()), ledger_impl_(ledger_impl) {
+    : ledger_client_(ledger_impl->ledger_client()),
+      ledger_impl_(ledger_impl),
+      options_(GetOptions()) {
   DCHECK(ledger_client_);
 }
 
 BATLedgerContext::BATLedgerContext(LedgerClient* ledger_client)
-    : ledger_client_(ledger_client) {
+    : ledger_client_(ledger_client), options_(GetOptions()) {
   DCHECK(ledger_client_);
 }
 
 BATLedgerContext::~BATLedgerContext() = default;
 
-using Component = BATLedgerContext::Component;
-using ComponentKey = BATLedgerContext::ComponentKey;
-
-ComponentKey::ComponentKey() : value_(g_next_component_key++) {}
+BATLedgerContext::Object::Object() = default;
 
-Component::Component(BATLedgerContext* context) : context_(context) {}
-
-Component::~Component() = default;
+BATLedgerContext::Object::~Object() = default;
 
 using LogStream = BATLedgerContext::LogStream;
 
-LogStream::LogStream(BATLedgerContext* context,
+LogStream::LogStream(base::WeakPtr<BATLedgerContext> context,
                      base::Location location,
                      LogLevel log_level)
-    : context_(context), location_(location), log_level_(log_level) {}
+    : context_(context), location_(location), log_level_(log_level) {
+  DCHECK(context);
+}
 
 LogStream::LogStream(LogStream&& other)
-    : context_(other.context_),
+    : context_(std::move(other.context_)),
```

Comment:
This should be `std::exchange(other.context_, nullptr)` in order for the destructor to work properly (unless the moved-from `base::WeakPtr<>` is guaranteed to refer to an object for which `*this == false`). Same goes for the move-assignment operator.

Reply:
I think that `base::WeakPtr<>` will invalidate itself when moved-from? I added a test, but let me know if I'm missing something.

Reply:
Just checked: this works due to `scoped_refptr<>`'s [move constructor](https://source.chromium.org/chromium/chromium/src/+/main:base/memory/scoped_refptr.h;l=206) and [copy-and-swap](https://source.chromium.org/chromium/chromium/src/+/main:base/memory/scoped_refptr.h;l=246-249) idiom implementation. It would be nice, though, if it was explicitly stated in `base::WeakPtr<>`'s documentation, since `scoped_refptr<const Flag>` is just an implementation detail of `base::internal::WeakReference`, which `base::WeakPtr<>` happens to use via `base::internal::WeakPtrBase`.